content/tutorials/exceptions.md

```diff
@@ -2,379 +2,697 @@
+and asynchronous exception is how they are thrown (via `throwIO` or
+`throwTo`). Unfortunately, there's no way to determine when catching
+an exception how it was thrown, making it difficult to live up to our
+goals above to never recover from an async exception. Fortunately, we
+have a workaround: use a different type for async exceptions!
+
+Using the hierarchical exception mechanism above, we have a new data
+type, `SomeAsyncException`, which is a child of `SomeException`. All
+exceptions which are thrown asynchronously must be a child of that
+exception type. And conversely, asynchronous exceptions must _not_ be
+thrown synchronously. The `UnliftIO.Exception` module has quite a few
+safeguards in place to ensure this happens; please see the "Async
+Exception Handling in Haskell" article above for the gory details.
+
+Upshot of all of this:
+
+* If you define your own exception type for asynchronous exceptions,
+  make it a child of `SomeAsyncException`. (Note: this is a pretty
+  unusal thing to do.)
+* If you define your own exception type for synchronous exceptions,
+  don't make it a child of `SomeAsyncException`.
+* The functions we'll mention below for cleaning up and recovering are
+  able to determine whether an exception is sync or async based on its
+  type.
+
+Cool? Awesome! That's quite enough backstory. Let's start covering
+usage of the API.
+
+## Throwing
+
+We're not going to talk about using async exceptions to kill other
+threads, since we're not talking about concurrent
+programming. Instead, please check out the [async
+library](/library/async) and the `race` and `cancel` functions it
+provides. Instead, we're going to focus on synchronous exception
+throwing.
+
+The most basic function for this is:
+
+```haskell
+throwIO :: (MonadIO m, Exception e) => e -> m a
+```
 
-__Exercise__ Write a function called `with2Files` which will open and
-guarantee the close of two files instead of just one. You can first
-implement in terms of `withFile`, and then implement without the help
-of `withFile`.
+Given any value which is an instance of `Exception`, you can throw it
+as a runtime exception for any monad which is a `MonadIO`
+instance. This works for built in exception types, as well as any you
+define yourself.
 
-__Solution__
+Sometimes you want to use synchronous exceptions but don't want to go
+through the overhead of defining your own exception type. In those
+cases, you can use the helper:
 
 ```haskell
-with2Files fp1 fp2 inner = mask $ \restore -> do
-    h1 <- openFile fp1
-    h2 <- openFile fp2 `onException` closeFile h1
-    restore (inner h1 h2) `finally` closeFile h2 `finally` closeFile h1
+throwString :: (MonadIO m, HasCallStack) => String -> m a
 ```
 
-The big takeaway you should get from this exercise: while masking
-prevent asynchronous exceptions from running, it does _nothing_ to
-prevent synchronous exceptions. You must still deal with those
-explicitly.
+`throwString` looks pretty similar to `error`:
 
```

Comment:
Probably it's useful to add here that it throws an exception of type `StringException` ?